### PR TITLE
add callout queue

### PR DESCRIPTION
**Pull request by Darge**
implements #45

### CODE REVIEW

File: callout.c [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+/*
+  Every event is inside one of NUMBER_OF_CALLOUT_BUCKETS buckets.
+  The buckets is a cyclic list, but we implement it as an array,
+  allowing us to access random elements in constant time.
+
+
+  TODO:
+  - long longs (nor int64_t) don't work: "Undefined reference to '__divdi3'"".
+    https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html, so I had
+    to change sbintime_t from int64_t to int.
```
Comment (cahirwpz) [inline callout.c]: Just link the kernel against libgcc - that should do the trick.

File: callout.c [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+
+void callout_init() {
+  memset(&ci, 0, sizeof ci);
+
+  for (int i = 0; i < NUMBER_OF_CALLOUT_BUCKETS; i++)
+    TAILQ_INIT(&ci.heads[i]);
+}
+
+void callout_setup(struct callout *handle, sbintime_t time, timeout_t fn,
```
Comment (cahirwpz) [inline callout.c]: Please be consistent `struct callout` == `callout_t`.

File: include/callout.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+typedef void (*timeout_t)(void *);
+typedef int sbintime_t;
+
+typedef struct callout {
+  TAILQ_ENTRY(callout) c_link;
+  sbintime_t c_time;  /* absolute time of the event */
+  timeout_t c_func;  /* function to call */
+  void    *c_arg;     /* function argument */
+  uint16_t c_flags;
+  unsigned int index; /* index of bucket this callout is assigned to */
```
Comment (cahirwpz) [inline include/callout.h]: Please use a type that explicitly states number of bits in word, i.e. uint32_t. Please make sure you use this convention through your whole code.

**Comment by cahirwpz** [comment]
Could you resolve conflicts and move this branch to my repo. I'd like to integrate it.

**Comment by Darge** [comment]
Branch added.
I created it from scratch, so that there are no merge commits - the history will be cleaner.

**Pull request closed without merging**